test/utils.test.ts

```diff
@@ -0,0 +1,205 @@
+import test from 'ava'
+import {parseURL} from 'ufo'
+import {decodeBase64Params, doPatternsMatchUrl, RemotePattern} from '../src/utils'
+
+const encodeUrlAndTransforms = (url: string, transforms: string) : string => {
+  const encodedUrl = Buffer.from(url).toString('base64')
+  const encodedTransforms = Buffer.from(transforms).toString('base64')
+  return `${encodedUrl}/${encodedTransforms}.ext`
+}
+
+test('decodeBase64Params: returns expected response if no url is present', (t) => {
+  const expectedResponse = {
+    error: 'Bad Request'
+  }
+  
+  const response = decodeBase64Params('')
+  
+  t.deepEqual(response, expectedResponse)
+})
+
+test('decodeBase64Params: returns expected response if no transform exists in path', (t) => {
+  const expectedResponse = {
+    error: 'Bad Request'
+  }
+  
+  const response = decodeBase64Params('fakeUrl/')
+  
+  t.deepEqual(response, expectedResponse)
+})
+
+// TODO: Can id and transform ever become falsy when base64 decoding if we're passing in a string?
```

Comment:
I couldn't see how `Buffer.from()` can fail in a way which would cause `id` and `transforms` to be empty. They'll `throw` if we don't pass a string through but then TypeScript should catch that in the first place. Let me know if you think an additional test is valuable for these.

Reply:
I think these will be falsy if the path is an empty string, which could happen if it's just the base path, i.e. `/_ipx/`

No big deal, but might be worth an extra test for completeness

Reply:
I'm trying to add an additional test but I still think that it won't become falsy when it attempts to base64 decode the string.

Here's the code:

```
export function decodeBase64Params (path:string) {
  const [url, transform] = path.split('/')
  if (!url || !transform) {
    return {
      error: 'Bad Request'
    }
  }
  const id = Buffer.from(url, 'base64').toString('utf8')
  // Strip the extension
  const transforms = Buffer.from(transform.split('.')[0], 'base64').toString(
    'utf8'
  )
  if (!id || !transforms) {
    return {
      error: 'Bad Request'
    }
  }
```

If the `url` or `transform` are empty strings then we'll fail at the first "Bad Request" check.
`Buffer.from` would throw an error if the `url` or the `transform.split('.')` are not strings but we don't `catch` here and TypeScript should protect us from not passing a string.

Can we ever get to the point where the 2nd check for `!id || !transforms` ever returns anything? 🤔 

@ascorbic - Am I missing something here? It could be my misunderstanding of how it works.

Reply:
Ah. You can get the `transforms` to fail if the `transform` parameter was just `.jpg` with no base64 encoded query parameters. Still not sure that we can do the same for `id`.